File: src/com/zephir/lang/ZephirBraceMatcher.java
// Copyright (c) 2014-2020 Zephir Team
//
// This source file is subject the MIT license, that is bundled with
// this package in the file LICENSE, and is available through the
// world-wide-web at the following url:
//
// https://github.com/zephir-lang/idea-plugin/blob/master/LICENSE

package com.zephir.lang;

import com.intellij.lang.BracePair;
import com.intellij.lang.PairedBraceMatcher;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import com.zephir.psi.ZephirTypes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ZephirBraceMatcher implements PairedBraceMatcher {
    private static final BracePair[] BRACE_PAIRS = {
            new BracePair(ZephirTypes.BRACKET_OPEN, ZephirTypes.BRACKET_CLOSE, true),
            new BracePair(ZephirTypes.SBRACKET_OPEN, ZephirTypes.SBRACKET_CLOSE, false),
            new BracePair(ZephirTypes.PARENTHESES_OPEN, ZephirTypes.PARENTHESES_CLOSE, false)
    };

    @Override
    public BracePair[] getPairs() {
        return BRACE_PAIRS;
    }

    @Override
    public boolean isPairedBracesAllowedBeforeType(@NotNull IElementType braceType, @Nullable IElementType contextType) {
        return true;
    }

    @Override
    public int getCodeConstructStart(PsiFile file, int openingBraceOffset) {
        return openingBraceOffset;
    }
}

File: src/main/kotlin/com/zephir/ide/typing/ZephirBraceMatcher.kt
// Copyright (c) 2014-2020 Zephir Team
//
// This source file is subject the MIT license, that is bundled with
// this package in the file LICENSE, and is available through the
// world-wide-web at the following url:
//
// https://github.com/zephir-lang/idea-plugin/blob/master/LICENSE

package com.zephir.ide.typing

import com.intellij.lang.BracePair
import com.intellij.lang.PairedBraceMatcher
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.TokenSet
import com.zephir.lang.core.psi.ZephirTypes.*
import com.zephir.lang.core.psi.ZEPHIR_COMMENTS

// TODO(serghei): re-visit this later to implement com.intellij.codeInsight.highlighting.PairedBraceMatcherAdapter
class ZephirBraceMatcher : PairedBraceMatcher {
    companion object {
        private val PAIRS: Array<BracePair> = arrayOf(
            BracePair(BRACKET_OPEN, BRACKET_CLOSE, true /* structural */),
            BracePair(SBRACKET_OPEN, SBRACKET_CLOSE, false),
            BracePair(PARENTHESES_OPEN, PARENTHESES_CLOSE, false),
            BracePair(LESS, GREATER, false)
        )

        private val InsertPairBraceBefore = TokenSet.orSet(
            ZEPHIR_COMMENTS,
            TokenSet.create(
                TokenType.WHITE_SPACE,
                DOTCOMMA,
                COMMA,
                PARENTHESES_CLOSE,
                SBRACKET_CLOSE,
                BRACKET_CLOSE,
                BRACKET_OPEN
            )
        )
    }

    override fun getCodeConstructStart(file: PsiFile?, openingBraceOffset: Int): Int = openingBraceOffset
    override fun getPairs(): Array<BracePair> = PAIRS
    override fun isPairedBracesAllowedBeforeType(lbraceType: IElementType, contextType: IElementType?) =
        contextType in InsertPairBraceBefore
}
